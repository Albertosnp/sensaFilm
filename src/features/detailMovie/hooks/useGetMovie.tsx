import { useEffect, useState } from 'react';
import { MovieDetailI } from '@/types';
import { getMovie } from '@/features/detailMovie/api/getMovie';

interface UseGetMovieProps {
  idMovie: string | undefined;
}

export const useGetMovie = ({ idMovie }: UseGetMovieProps) => {
  const [movie, setMovie] = useState<MovieDetailI[] | null>(null);
  const [loading, setLoading] = useState(false);

  useEffect(() => {
    if (!idMovie) return;

    let cancelled = true;
    setLoading(true);

    getMovie(idMovie)
      .then((res) => {
        if (!cancelled) return;

        setMovie(movie);
        setLoading(false);
      })
      .catch((err) => console.error(err));

    return () => {
      cancelled = false;
    };
  }, [idMovie]);

  return { movie, loading } as const;
};
