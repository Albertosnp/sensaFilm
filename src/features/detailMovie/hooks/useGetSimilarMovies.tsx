import { useEffect, useState } from 'react';
import { MovieI } from '@/types';
import { getSimilarMovies } from '../api/getSimilarMovies';

interface UseGetMovieProps {
  idMovie: string | undefined;
}

export const useGetSimilarMovies = ({ idMovie }: UseGetMovieProps) => {
  const [movies, setMovies] = useState<MovieI[] | null>(null);
  const [loading, setLoading] = useState(false);

  useEffect(() => {
    if (!idMovie) return;

    let cancelled = true;
    setLoading(true);

    getSimilarMovies(idMovie)
      .then((res) => {
        if (!cancelled) return;

        const similarMovies = res?.results;
        setMovies(similarMovies);
        setLoading(false);
      })
      .catch((err) => console.error(err));

    return () => {
      cancelled = false;
    };
  }, [idMovie]);

  return { movies, loading } as const;
};
