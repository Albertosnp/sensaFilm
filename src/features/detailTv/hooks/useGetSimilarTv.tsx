import { useEffect, useState } from 'react';
import { TvProgramI } from '@/types';
import { getSimilarTv } from '../api/getSimilarTv';

interface UseGetMovieProps {
  idTv: string | undefined;
}

export const useGetSimilarTv = ({ idTv }: UseGetMovieProps) => {
  const [tvPrograms, setTvPrograms] = useState<TvProgramI[] | null>(null);
  const [loading, setLoading] = useState(false);

  useEffect(() => {
    if (!idTv) return;

    let cancelled = true;
    setLoading(true);

    getSimilarTv(idTv)
      .then((res) => {
        if (!cancelled) return;

        const programs = res?.results;

        setTvPrograms(programs);
        setLoading(false);
      })
      .catch((err) => console.error(err));

    return () => {
      cancelled = false;
    };
  }, [idTv]);

  return { tvPrograms, loading } as const;
};
