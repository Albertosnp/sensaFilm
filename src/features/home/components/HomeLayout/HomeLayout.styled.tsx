import styled from '@emotion/styled';
import breakpoints from '@/assets/styles/breakpoints';
import { rem } from '@/assets/styles/_mixins';

const HomeLayoutStyled = styled.div`
  padding: 5%;
  .HomeLayout {
    &__grid {
      display: grid;
      grid-template-columns: repeat(auto-fit, minmax(250px, 1fr));
      column-gap: 30px;
      row-gap: 50px;
    }
  }
  @media (min-width: ${rem(breakpoints.tablet)}) {
  }
  @media (min-width: ${rem(breakpoints.tablet_wide)}) {
  }
  @media (min-width: ${rem(breakpoints.desktop)}) {
  }
`;
export default HomeLayoutStyled;
