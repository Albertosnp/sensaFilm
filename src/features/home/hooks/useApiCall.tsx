import { useEffect, useState } from 'react';

export const useApiCall = () => {
  const [result, setResult] = useState();
  const [loading, setLoading] = useState(false);

  useEffect(() => {
    let cancelled = true;
    setLoading(true);

    //API call

    return () => {
      cancelled = false;
    };
  }, []);

  return [result, loading] as const;
};
