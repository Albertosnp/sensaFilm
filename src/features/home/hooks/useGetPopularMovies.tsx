import { useEffect, useState } from 'react';
import { getPopularMovies } from '../api/getPopularMovies';
import { MovieI } from '@/types';

export const useGetPopularMovies = () => {
  const [popularMovies, setPopularMovies] = useState<MovieI[] | null>(null);
  const [loading, setLoading] = useState(false);

  useEffect(() => {
    let cancelled = true;
    setLoading(true);

    getPopularMovies()
      .then((res) => {
        if (!cancelled) return;

        const movies = res?.results;
        setTimeout(() => {
          setPopularMovies(movies);

          setLoading(false);
        }, 2000);
      })
      .catch((err) => console.error(err));

    return () => {
      cancelled = false;
    };
  }, []);

  return { popularMovies, loading } as const;
};
