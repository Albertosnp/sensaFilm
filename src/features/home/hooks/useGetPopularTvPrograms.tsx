import { useEffect, useState } from 'react';
import { getPopularTvPrograms } from '../api/getPopularTvPrograms';
import { TvProgramI } from '@/types';

export const useGetPopularTvPrograms = () => {
  const [popularTvPrograms, setPopularTvPrograms] = useState<
    TvProgramI[] | null
  >(null);
  const [loading, setLoading] = useState(false);

  useEffect(() => {
    let cancelled = true;
    setLoading(true);

    getPopularTvPrograms()
      .then((res) => {
        if (!cancelled) return;

        const tvPrograms = res?.results;

        setPopularTvPrograms(tvPrograms);
        setLoading(false);
      })
      .catch((err) => console.error(err));

    return () => {
      cancelled = false;
    };
  }, []);

  return { popularTvPrograms, loading } as const;
};
