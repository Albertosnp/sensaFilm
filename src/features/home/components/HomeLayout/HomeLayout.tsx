import { GenericCard, SkeletonCard } from '@/components';
import { useCallback, useState } from 'react';
import { useGetPopularMovies } from '../../hooks/useGetPopularMovies';
import { useGetPopularTvPrograms } from '../../hooks/useGetPopularTvPrograms';
import { HomeHeader } from '../HomeHeader/HomeHeader';

import HomeLayoutStyled from './HomeLayout.styled';

export const HomeLayout = () => {
  const [toggleCategory, setToggleCategory] = useState(true);
  const { popularMovies } = useGetPopularMovies();
  const { popularTvPrograms } = useGetPopularTvPrograms();

  const handleToggleCategory = useCallback(() => {
    setToggleCategory((prev) => !prev);
  }, []);

  return (
    <HomeLayoutStyled className="HomeLayout">
      <HomeHeader
        toggleCategory={toggleCategory}
        handleToggleCategory={handleToggleCategory}
      />

      {popularMovies && (
        <section className="HomeLayout__grid">
          {toggleCategory &&
            popularMovies?.map(({ id, poster_path, title, vote_average }) => (
              <GenericCard
                key={`${id}-movie`}
                id={id}
                title={title}
                image={poster_path ?? ''}
                vote_average={vote_average}
                path="detailMovie"
              />
            ))}
          {!toggleCategory &&
            popularTvPrograms?.map(
              ({ id, poster_path, name, vote_average }) => (
                <GenericCard
                  key={`${id}-tv`}
                  id={id}
                  title={name}
                  image={poster_path ?? ''}
                  vote_average={vote_average}
                  path="detailTv"
                />
              )
            )}
        </section>
      )}
      {!popularMovies && (
        <section className="HomeLayout__grid">
          {[0, 0, 0, 0, 0, 0].map((skeleton, index) => (
            <SkeletonCard key={`${index}-skeleton`} />
          ))}
        </section>
      )}
    </HomeLayoutStyled>
  );
};
