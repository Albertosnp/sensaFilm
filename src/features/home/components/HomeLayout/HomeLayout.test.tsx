import { render, screen } from "@testing-library/react";
import { HomeLayout } from "./HomeLayout";
import { createRoot } from "react-dom/client";
import { vi } from "vitest";
import { useGetPopularTvPrograms } from "../../hooks/useGetPopularTvPrograms";
import { useGetPopularMovies } from "../../hooks/useGetPopularMovies";
import React from "react";

vi.mock("../../hooks/useGetPopularMovies", () => {
  return {
    ...vi.importMock("../../hooks/useGetPopularMovies"),
    useGetPopularMovies: vi.fn(),
  };
});

vi.mock("../../hooks/useGetPopularTvPrograms", () => {
  return {
    ...vi.importMock("../../hooks/useGetPopularTvPrograms"),
    useGetPopularTvPrograms: vi.fn(),
  };
});

vi.mock("react-router-dom", async () => {
  return {
    ...vi.importMock("react-router-dom"),
    Link: ({ children, to }: { children: JSX.Element; to: string }) =>
      React.createElement("a", { href: to }, children),
  };
});

const popularTvPrograms = [
  {
    backdrop_path: "/jWXrQstj7p3Wl5MfYWY6IHqRpDb.jpg",
    first_air_date: "1952-12-26",
    genre_ids: [10763],
    id: 94722,
    name: "Tagesschau",
    origin_country: ["DE"],
    original_language: "de",
    original_name: "Tagesschau",
    overview: "",
    popularity: 4082.266,
    poster_path: "/7dFZJ2ZJJdcmkp05B9NWlqTJ5tq.jpg",
    vote_average: 7.5,
    vote_count: 130,
  },
  {
    backdrop_path: "/t2rAdgjSh0WYbXzdOB5zTDqzdCI.jpg",
    first_air_date: "2022-11-02",
    genre_ids: [18],
    id: 213713,
    name: "Faltu",
    origin_country: ["IN"],
    original_language: "hi",
    original_name: "Faltu",
    overview: "",
    popularity: 3157.698,
    poster_path: "/lgyFuoXs7GvKJN0mNm7z7OMOFuZ.jpg",
    vote_average: 7,
    vote_count: 134,
  },
];
const popularMovies = [
  {
    adult: false,
    backdrop_path: "/4HodYYKEIsGOdinkGi2Ucz6X9i0.jpg",
    genre_ids: [16, 28, 12],
    id: 569094,
    original_language: "en",
    original_title: "Spider-Man: Across the Spider-Verse",
    overview:
      "Miles Morales regresa para una aventura épica que transportará al amigable vecino de Brooklyn Spider-Man a través del Multiverso para unir fuerzas con Gwen Stacy y un nuevo equipo de Spider-People, y enfrentarse así a un villano mucho más poderoso que cualquier cosa que hayan conocido antes.",
    popularity: 2044.25,
    poster_path: "/37WcNMgNOMxdhT87MFl7tq7FM1.jpg",
    release_date: "2023-05-31",
    title: "Spider-Man: Cruzando el Multiverso",
    video: false,
    vote_average: 8.4,
    vote_count: 3775,
  },
  {
    adult: false,
    backdrop_path: "/nHf61UzkfFno5X1ofIhugCPus2R.jpg",
    genre_ids: [35, 12, 14],
    id: 346698,
    original_language: "en",
    original_title: "Barbie",
    overview:
      "Barbie lleva una vida ideal en Barbielandia, allí todo es perfecto, con chupifiestas llenas de música y color, y todos los días son el mejor día. Claro que Barbie se hace algunas preguntas, cuestiones bastante incómodas que no encajan con el mundo idílico en el que ella y las demás Barbies viven. Cuando Barbie se dé cuenta de que es capaz de apoyar los talones en el suelo, y tener los pies planos, decidirá calzarse unos zapatos sin tacones y viajar hasta el mundo real.",
    popularity: 1659.364,
    poster_path: "/fNtqD4BTFj0Bgo9lyoAtmNFzxHN.jpg",
    release_date: "2023-07-19",
    title: "Barbie",
    video: false,
    vote_average: 7.4,
    vote_count: 3404,
  },
  {
    adult: false,
    backdrop_path: "/2vFuG6bWGyQUzYS9d69E5l85nIz.jpg",
    genre_ids: [28, 12, 878],
    id: 667538,
    original_language: "en",
    original_title: "Transformers: Rise of the Beasts",
    overview:
      "Cuando surge una nueva amenaza capaz de destruir todo el planeta, Optimus Prime y los Autobots deben unirse a una poderosa facción conocida como los Maximals. Con el destino de la humanidad en juego, los humanos Noah y Elena harán lo que sea necesario para ayudar a los Transformers mientras se involucran en la batalla final para salvar la Tierra.",
    popularity: 1646.07,
    poster_path: "/8HBYMQkF6cFpupgRDhsrrKskXOE.jpg",
    release_date: "2023-06-06",
    title: "Transformers: El despertar de las bestias",
    video: false,
    vote_average: 7.5,
    vote_count: 2947,
  },
];

describe("HomeLayout", () => {
  const defaultProps = {};
  const getComponent = (props = defaultProps) => <HomeLayout {...props} />;
  const setup = (props = defaultProps) => render(getComponent(props));

  beforeEach(() => {
    // @ts-ignore
    useGetPopularMovies.mockReturnValue({ popularMovies });
    // @ts-ignore
    useGetPopularTvPrograms.mockReturnValue({ popularTvPrograms });
  });

  it("Should render without crashing", () => {
    const div = document.createElement("div");
    createRoot(div).render(getComponent());
    createRoot(div).unmount();
  });

  it("Should match snapshot with data", () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it("Should change data when click input switch", () => {
    setup();
    expect(screen.queryByText("Tagesschau")).not.toBeInTheDocument();
    screen.getByRole("checkbox").click();
    expect(screen.queryByText("Tagesschau")).toBeInTheDocument();
  });

  it("Should render skeleton without data", () => {
    // @ts-ignore
    useGetPopularMovies.mockReturnValue({});
    // @ts-ignore
    useGetPopularTvPrograms.mockReturnValue({});
    setup();
    screen.getByTestId("skeleton");
  });
});
