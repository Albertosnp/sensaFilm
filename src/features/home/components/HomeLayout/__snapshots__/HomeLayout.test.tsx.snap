// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HomeLayout > Should match snapshot with data 1`] = `
<div>
  <div
    class="HomeLayout css-n9kgpy"
  >
    <div
      class="HomeHeader css-hn9zvr"
    >
      <h3
        class="CategoryTitle css-ncm7rz"
      >
        Pel√≠culas destacadas
      </h3>
      <div
        class="HomeHeader__buttons"
      >
        <img
          class="HomeHeader__buttons__img"
          src="/src/assets/img/video.png"
        />
        <div
          class="inputSwitch css-1rzg66u"
        >
          <label
            class="inputSwitch__label"
          >
            <input
              class="inputSwitch__input"
              id="-"
              name=""
              type="checkbox"
              value=""
            />
            <span
              class="inputSwitch__label__icon"
            />
          </label>
        </div>
        <img
          class="HomeHeader__buttons__img"
          src="/src/assets/img/television.png"
        />
      </div>
    </div>
    <section
      class="HomeLayout__grid"
    >
      <figure
        class="GenericCard css-1sbhet6"
      >
        <a
          href="detailMovie/569094"
        >
          <img
            alt="imagen-picture"
            class="GenericCard__image"
            src="https://image.tmdb.org/t/p/w500//37WcNMgNOMxdhT87MFl7tq7FM1.jpg"
          />
          <div
            class="GenericCard__footer"
          >
            <p
              class="GenericCard__footer__title"
            >
              Spider-Man: Cruzando el Multiverso
            </p>
            <span
              class="GenericCard__footer__votes"
            >
              8.4
              /10
            </span>
          </div>
        </a>
      </figure>
      <figure
        class="GenericCard css-1sbhet6"
      >
        <a
          href="detailMovie/346698"
        >
          <img
            alt="imagen-picture"
            class="GenericCard__image"
            src="https://image.tmdb.org/t/p/w500//fNtqD4BTFj0Bgo9lyoAtmNFzxHN.jpg"
          />
          <div
            class="GenericCard__footer"
          >
            <p
              class="GenericCard__footer__title"
            >
              Barbie
            </p>
            <span
              class="GenericCard__footer__votes"
            >
              7.4
              /10
            </span>
          </div>
        </a>
      </figure>
      <figure
        class="GenericCard css-1sbhet6"
      >
        <a
          href="detailMovie/667538"
        >
          <img
            alt="imagen-picture"
            class="GenericCard__image"
            src="https://image.tmdb.org/t/p/w500//8HBYMQkF6cFpupgRDhsrrKskXOE.jpg"
          />
          <div
            class="GenericCard__footer"
          >
            <p
              class="GenericCard__footer__title"
            >
              Transformers: El despertar de las bestias
            </p>
            <span
              class="GenericCard__footer__votes"
            >
              7.5
              /10
            </span>
          </div>
        </a>
      </figure>
    </section>
  </div>
</div>
`;
